src/PageContent/AssignVolunteers/AssignVolunteersController.js

```diff
@@ -1,145 +1,92 @@
-                    }
-                ]
-            },
-            description: {
-                foodItems: [
-                    {
-                        food: 'Baked beans',
-                        quantity: 15,
-                        unit: 'lb'  // Enums.FoodUnit
-                    },
-                    {
-                        food: 'Bread',
-                        quantity: 4,
-                        unit: 'loaves'  // Enums.FoodUnit
-                    },
-                ]
-            }
-        });
-        deliveryList.push({
-            date: '2018-02-26',
-            startTime: '14:00',
-            endTime: '17:00',
-            donatingAgency: {
-                agency: 'Local Point',
-                address: 'Test Address',
-                primaryContact: {
-                    name: 'Alice',
-                    phone: 7739939922
-                }
-            },
-            receivingAgency: {
-                agency: 'Union Gospel Shelter',
-                primaryContact: {
-                    name: 'Bob',
-                    email: 'bob@uniongospel.org',
-                    phone: 1237894560
+    componentDidMount() {
+        let genAccountPromise = (deliveryId, accountId) => (new Promise(async (resolve, reject) => {
+            let snap = await db.ref(`accounts/${accountId}`).once('value');
+            resolve({deliveryId: deliveryId,
+                     account: snap.val()});
+        }));
+
+        let genSupplementaryPromise = (pathPrefix, deliveryId, accountId) => (new Promise(async (resolve, reject) =>
+            resolve((await db.ref(`${pathPrefix}/${accountId}`).once('value')).val())));
+        
+        let genDeliveryListener = (deliveryId) => {
+            db.ref(`deliveries/${deliveryId}`).on('value', async (snap) => {
+                let delivery = snap.val();
+
+                // kick all promises off
+                let donatingAgencyPromise = genSupplementaryPromise('donating_agencies', deliveryId, delivery.donatingAgency);
+                let donatingAgencyContactPromise = genAccountPromise(deliveryId, delivery.daContact);
+                let receivingAgencyPromise = genAccountPromise(deliveryId, delivery.receivingAgency);
+                // collect them
+                let donatingAgency = await donatingAgencyPromise;
+                let donatingAgencyContact = await donatingAgencyContactPromise;
+                let receivingAgency = await receivingAgencyPromise;
+
+                delivery.donatingAgency = donatingAgency;
+                delivery.daContact = donatingAgencyContact.account;
+                delivery.receivingAgency =  receivingAgency.account.name;
+                delivery.updatedTime = moment().valueOf();
+
+                let updater = {}
+                updater[deliveryId] = {$set: delivery};
+                this.setState(prevState => ({deliveries: update(prevState.deliveries, updater)})); // does this handle async setState?
```

Comment:
yes... In fact I think the way you've described is cleaner. My goal was to never mutate state, instead copying the old deliveries and appending the new addition to the copy rather than the original. But I think it's actually fine to do it the way that you are, since the functional version of `setState` executes in sequence. My only concern is that we're technicaly directly modifying the state (since prevState.deliveries is a reference), but I think it's ok. I'll update this to follow your pattern.